lib/rmg/RiverPlacer.cpp

```diff
@@ -0,0 +1,403 @@
+#include "Functions.h"
+#include "CMapGenerator.h"
+#include "RmgMap.h"
+#include "../mapping/CMap.h"
+#include "../mapping/CMapEditManager.h"
+#include "../mapObjects/CObjectClassesHandler.h"
+#include "RmgPath.h"
+#include "ObjectManager.h"
+#include "ObstaclePlacer.h"
+#include "WaterProxy.h"
+#include "RoadPlacer.h"
+
+const int RIVER_DELTA_ID = 143;
+const int RIVER_DELTA_SUBTYPE = 0;
+const std::map<std::string, std::string> RIVER_DELTA_TEMPLATE_NAME
+{
+	{RIVER_NAMES[1], "clrdelt"},
+	{RIVER_NAMES[2], "icedelt"},
+	{RIVER_NAMES[3], "muddelt"},
+	{RIVER_NAMES[4], "lavdelt"}
+};
+
+const std::array<std::array<int, 25>, 4> deltaTemplates
+{
+	//0 - must be on ground
+	//1 - delta entry
+	//2 - must be on water
+	//3 - anything
+	//4 - prohibit river placement
+	//5 - must be on ground + position
+	//6 - must be on water + position
+	std::array<int, 25>{
+		3, 4, 3, 4, 3,
+		3, 4, 1, 4, 3,
+		3, 0, 0, 0, 3,
+		3, 0, 0, 0, 3,
+		3, 2, 2, 6, 3
+	},
+	std::array<int, 25>{
+		3, 2, 2, 2, 3,
+		3, 0, 0, 0, 3,
+		3, 0, 0, 5, 3,
+		3, 4, 1, 4, 3,
+		3, 4, 3, 4, 3
+	},
+	std::array<int, 25>{
+		3, 3, 3, 3, 3,
+		4, 4, 0, 0, 2,
+		3, 1, 0, 0, 2,
+		4, 4, 0, 0, 6,
+		3, 3, 3, 3, 3
+	},
+	std::array<int, 25>	{
+		3, 3, 3, 3, 3,
+		2, 0, 0, 4, 4,
+		2, 0, 0, 1, 3,
+		2, 0, 5, 4, 4,
+		3, 3, 3, 3, 3
+	}
+};
+
+void RiverPlacer::process()
+{
+	preprocess();
+	for(auto & t : riverNodes)
+		connectRiver(t);
+	
+	if(!rivers.empty())
+		drawRivers();
+}
+
+void RiverPlacer::init()
+{
+	DEPENDENCY_ALL(WaterProxy);
+	DEPENDENCY(ObjectManager);
+	DEPENDENCY(ObstaclePlacer);
+}
+
+void RiverPlacer::drawRivers()
+{
+	map.getEditManager()->getTerrainSelection().setSelection(rivers.getTilesVector());
+	map.getEditManager()->drawRiver(Terrain::Manager::getInfo(zone.getTerrainType()).river, &generator.rand);
+}
+
+char RiverPlacer::dump(const int3 & t)
+{
+	if(riverNodes.count(t))
+		return '@';
+	if(rivers.contains(t))
+		return '~';
+	if(sink.contains(t))
+		return '2';
+	if(source.contains(t))
+		return '1';
+	if(zone.area().contains(t))
+		return ' ';
+	return '?';
+}
+
+void RiverPlacer::addRiverNode(const int3 & node)
+{
+	assert(zone.area().contains(node));
+	riverNodes.insert(node);
+}
+
+rmg::Area & RiverPlacer::riverSource()
+{
+	return source;
+}
+
+rmg::Area & RiverPlacer::riverSink()
+{
+	return sink;
+}
+
+rmg::Area & RiverPlacer::riverProhibit()
+{
+	return prohibit;
+}
+
+void RiverPlacer::prepareHeightmap()
+{
+	rmg::Area roads;
+	if(auto * m = zone.getModificator<RoadPlacer>())
+	{
+		roads.unite(m->getRoads());
+	}
+	
+	for(auto & t : zone.area().getTilesVector())
+	{
+		heightMap[t] = generator.rand.nextInt(5);
+		
+		if(roads.contains(t))
+			heightMap[t] += 30.f;
+		
+		if(zone.areaUsed().contains(t))
+			heightMap[t] += 1000.f;
+	}
+	
+	//make grid
+	for(int j = 0; j < map.map().height; j += 2)
+	{
+		for(int i = 0; i < map.map().width; i += 2)
+		{
+			int3 t{i, j, zone.getPos().z};
+			if(zone.area().contains(t))
+				heightMap[t] += 10.f;
```

Comment:
Just an idea. What if we add some random noise here to let rivers be not too straight?

Reply:
Same idea goes for zone borders. I don`t like straight zone borders in maps generated by RMGs in Heroes 5 - 7.

Reply:
Borders are a different question. Not sure even if this simple idea with noise will even work.

Reply:
I tried to make it but result was strange. Probably, rivers need some efforts to look nice

Reply:
Zone borders could be fractalized - I have plans to do it later